File: Tests/MUAudioSessionManagerTests.m
#import <XCTest/XCTest.h>
#import <AVFoundation/AVFoundation.h>
#import <objc/runtime.h>

// Import MumbleKit if available, otherwise use stubs
#if __has_include(<MumbleKit/MKAudio.h>)
#import <MumbleKit/MKAudio.h>
#else
// Minimal stubs to allow the tests to compile in environments where
// the real MumbleKit headers are not available (such as CI runners).
@interface MKAudio : NSObject
@property (nonatomic) BOOL isRunning;
- (void)start;
- (void)stop;
- (void)restart;
@end

@implementation MKAudio
- (void)start {}
- (void)stop {}
- (void)restart {}
@end
#endif

// Mock MKAudio for testing that tracks method calls
@interface MockMKAudio : MKAudio
@property (nonatomic) NSInteger startCallCount;
@property (nonatomic) NSInteger stopCallCount;
@property (nonatomic) NSInteger restartCallCount;
@property (nonatomic) BOOL mockIsRunning;
@end

@implementation MockMKAudio

- (BOOL)isRunning {
    return self.mockIsRunning;
}

- (void)start {
    self.startCallCount += 1;
    self.mockIsRunning = YES;
}

- (void)stop {
    self.stopCallCount += 1;
    self.mockIsRunning = NO;
}

- (void)restart {
    self.restartCallCount += 1;
}

@end

@interface MUAudioSessionManagerTests : XCTestCase
@property (nonatomic, strong) MockMKAudio *mockAudio;
@property (nonatomic, strong) NSUserDefaults *mockDefaults;
@property (nonatomic, strong) id sessionManager;
@end

@implementation MUAudioSessionManagerTests

- (void)setUp {
    [super setUp];
    
    // Use a fresh UserDefaults suite for each test
    NSString *suiteName = [NSString stringWithFormat:@"test.suite.%@", [[NSUUID UUID] UUIDString]];
    self.mockDefaults = [[NSUserDefaults alloc] initWithSuiteName:suiteName];
    self.mockAudio = [[MockMKAudio alloc] init];
    
    // Access the Swift singleton via Objective-C runtime
    Class managerClass = NSClassFromString(@"Mumble.MUAudioSessionManager");
    if (managerClass) {
        self.sessionManager = [managerClass performSelector:@selector(shared)];
    }
}

- (void)tearDown {
    // Clean up the test defaults suite
    [self.mockDefaults removePersistentDomainForName:self.mockDefaults.name];
    self.mockDefaults = nil;
    self.mockAudio = nil;
    self.sessionManager = nil;
    
    [super tearDown];
}
    

// MARK: - bind(to:defaults:) Tests

- (void)testBindAppliesPlaybackPreferences {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: Speaker mode preference is enabled
    [self.mockDefaults setBool:YES forKey:@"AudioSpeakerPhoneMode"];
    
    // When: Binding to MKAudio
    SEL bindSelector = NSSelectorFromString(@"bindToMumbleKitAudio:defaults:");
    if ([self.sessionManager respondsToSelector:bindSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:bindSelector withObject:self.mockAudio withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    
    // Then: Should complete without error
    XCTAssertNotNil(self.sessionManager);
}

- (void)testBindAppliesDefaultSpeakerModeWhenNotSet {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: No speaker mode preference set (defaults to false)
    
    // When: Binding to MKAudio
    SEL bindSelector = NSSelectorFromString(@"bindToMumbleKitAudio:defaults:");
    if ([self.sessionManager respondsToSelector:bindSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:bindSelector withObject:self.mockAudio withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    
    // Then: Should complete without error
    XCTAssertNotNil(self.sessionManager);
}

// MARK: - refreshPlaybackChain() Tests

- (void)testRefreshPlaybackChainRestartsRunningAudio {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: Audio is bound and running
    self.mockAudio.mockIsRunning = YES;
    SEL bindSelector = NSSelectorFromString(@"bindToMumbleKitAudio:defaults:");
    if ([self.sessionManager respondsToSelector:bindSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:bindSelector withObject:self.mockAudio withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    
    // When: Refreshing the playback chain
    SEL refreshSelector = NSSelectorFromString(@"refreshPlaybackChain");
    if ([self.sessionManager respondsToSelector:refreshSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:refreshSelector];
#pragma clang diagnostic pop
    }
    
    // Then: Audio should be restarted
    XCTAssertEqual(self.mockAudio.restartCallCount, 1);
}

- (void)testRefreshPlaybackChainStartsStoppedAudio {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: Audio is bound but not running
    self.mockAudio.mockIsRunning = NO;
    SEL bindSelector = NSSelectorFromString(@"bindToMumbleKitAudio:defaults:");
    if ([self.sessionManager respondsToSelector:bindSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:bindSelector withObject:self.mockAudio withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    
    // Reset call counts after bind
    self.mockAudio.startCallCount = 0;
    
    // When: Refreshing the playback chain
    SEL refreshSelector = NSSelectorFromString(@"refreshPlaybackChain");
    if ([self.sessionManager respondsToSelector:refreshSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:refreshSelector];
#pragma clang diagnostic pop
    }
    
    // Then: Audio should be started
    XCTAssertEqual(self.mockAudio.startCallCount, 1);
}

// MARK: - handleRouteChange(reasonValue:defaults:) Tests

- (void)testHandleRouteChangeWithNewDeviceRestartsAudio {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: Audio is bound and running
    self.mockAudio.mockIsRunning = YES;
    SEL bindSelector = NSSelectorFromString(@"bindToMumbleKitAudio:defaults:");
    if ([self.sessionManager respondsToSelector:bindSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:bindSelector withObject:self.mockAudio withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    self.mockAudio.restartCallCount = 0;
    
    // When: Route changes due to new device
    AVAudioSessionRouteChangeReason reason = AVAudioSessionRouteChangeReasonNewDeviceAvailable;
    SEL handleSelector = NSSelectorFromString(@"handleRouteChangeWithReason:defaults:");
    if ([self.sessionManager respondsToSelector:handleSelector]) {
        NSMethodSignature *signature = [self.sessionManager methodSignatureForSelector:handleSelector];
        NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:signature];
        [invocation setSelector:handleSelector];
        [invocation setTarget:self.sessionManager];
        
        NSUInteger reasonValue = (NSUInteger)reason;
        [invocation setArgument:&reasonValue atIndex:2];
        [invocation setArgument:&_mockDefaults atIndex:3];
        [invocation invoke];
    }
    
    // Then: Audio should be restarted
    XCTAssertEqual(self.mockAudio.restartCallCount, 1);
}

- (void)testHandleRouteChangeWithOldDeviceUnavailableRestartsAudio {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: Audio is bound and running
    self.mockAudio.mockIsRunning = YES;
    SEL bindSelector = NSSelectorFromString(@"bindToMumbleKitAudio:defaults:");
    if ([self.sessionManager respondsToSelector:bindSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:bindSelector withObject:self.mockAudio withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    self.mockAudio.restartCallCount = 0;
    
    // When: Route changes due to device removal
    AVAudioSessionRouteChangeReason reason = AVAudioSessionRouteChangeReasonOldDeviceUnavailable;
    SEL handleSelector = NSSelectorFromString(@"handleRouteChangeWithReason:defaults:");
    if ([self.sessionManager respondsToSelector:handleSelector]) {
        NSMethodSignature *signature = [self.sessionManager methodSignatureForSelector:handleSelector];
        NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:signature];
        [invocation setSelector:handleSelector];
        [invocation setTarget:self.sessionManager];
        
        NSUInteger reasonValue = (NSUInteger)reason;
        [invocation setArgument:&reasonValue atIndex:2];
        [invocation setArgument:&_mockDefaults atIndex:3];
        [invocation invoke];
    }
    
    // Then: Audio should be restarted
    XCTAssertEqual(self.mockAudio.restartCallCount, 1);
}

- (void)testHandleRouteChangeWithCategoryChangeRestartsAudio {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: Audio is bound and running
    self.mockAudio.mockIsRunning = YES;
    SEL bindSelector = NSSelectorFromString(@"bindToMumbleKitAudio:defaults:");
    if ([self.sessionManager respondsToSelector:bindSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:bindSelector withObject:self.mockAudio withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    self.mockAudio.restartCallCount = 0;
    
    // When: Route changes due to category change
    AVAudioSessionRouteChangeReason reason = AVAudioSessionRouteChangeReasonCategoryChange;
    SEL handleSelector = NSSelectorFromString(@"handleRouteChangeWithReason:defaults:");
    if ([self.sessionManager respondsToSelector:handleSelector]) {
        NSMethodSignature *signature = [self.sessionManager methodSignatureForSelector:handleSelector];
        NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:signature];
        [invocation setSelector:handleSelector];
        [invocation setTarget:self.sessionManager];
        
        NSUInteger reasonValue = (NSUInteger)reason;
        [invocation setArgument:&reasonValue atIndex:2];
        [invocation setArgument:&_mockDefaults atIndex:3];
        [invocation invoke];
    }
    
    // Then: Audio should be restarted
    XCTAssertEqual(self.mockAudio.restartCallCount, 1);
}

- (void)testHandleRouteChangeWithOverrideRestartsAudio {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: Audio is bound and running
    self.mockAudio.mockIsRunning = YES;
    SEL bindSelector = NSSelectorFromString(@"bindToMumbleKitAudio:defaults:");
    if ([self.sessionManager respondsToSelector:bindSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:bindSelector withObject:self.mockAudio withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    self.mockAudio.restartCallCount = 0;
    
    // When: Route changes due to override
    AVAudioSessionRouteChangeReason reason = AVAudioSessionRouteChangeReasonOverride;
    SEL handleSelector = NSSelectorFromString(@"handleRouteChangeWithReason:defaults:");
    if ([self.sessionManager respondsToSelector:handleSelector]) {
        NSMethodSignature *signature = [self.sessionManager methodSignatureForSelector:handleSelector];
        NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:signature];
        [invocation setSelector:handleSelector];
        [invocation setTarget:self.sessionManager];
        
        NSUInteger reasonValue = (NSUInteger)reason;
        [invocation setArgument:&reasonValue atIndex:2];
        [invocation setArgument:&_mockDefaults atIndex:3];
        [invocation invoke];
    }
    
    // Then: Audio should be restarted
    XCTAssertEqual(self.mockAudio.restartCallCount, 1);
}

- (void)testHandleRouteChangeWithWakeFromSleepRestartsAudio {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: Audio is bound and running
    self.mockAudio.mockIsRunning = YES;
    SEL bindSelector = NSSelectorFromString(@"bindToMumbleKitAudio:defaults:");
    if ([self.sessionManager respondsToSelector:bindSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:bindSelector withObject:self.mockAudio withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    self.mockAudio.restartCallCount = 0;
    
    // When: Route changes due to wake from sleep
    AVAudioSessionRouteChangeReason reason = AVAudioSessionRouteChangeReasonWakeFromSleep;
    SEL handleSelector = NSSelectorFromString(@"handleRouteChangeWithReason:defaults:");
    if ([self.sessionManager respondsToSelector:handleSelector]) {
        NSMethodSignature *signature = [self.sessionManager methodSignatureForSelector:handleSelector];
        NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:signature];
        [invocation setSelector:handleSelector];
        [invocation setTarget:self.sessionManager];
        
        NSUInteger reasonValue = (NSUInteger)reason;
        [invocation setArgument:&reasonValue atIndex:2];
        [invocation setArgument:&_mockDefaults atIndex:3];
        [invocation invoke];
    }
    
    // Then: Audio should be restarted
    XCTAssertEqual(self.mockAudio.restartCallCount, 1);
}

- (void)testHandleRouteChangeWithUnknownReasonDoesNotRestartAudio {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: Audio is bound and running
    self.mockAudio.mockIsRunning = YES;
    SEL bindSelector = NSSelectorFromString(@"bindToMumbleKitAudio:defaults:");
    if ([self.sessionManager respondsToSelector:bindSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:bindSelector withObject:self.mockAudio withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    self.mockAudio.restartCallCount = 0;
    
    // When: Route changes for an unknown reason
    AVAudioSessionRouteChangeReason reason = AVAudioSessionRouteChangeReasonUnknown;
    SEL handleSelector = NSSelectorFromString(@"handleRouteChangeWithReason:defaults:");
    if ([self.sessionManager respondsToSelector:handleSelector]) {
        NSMethodSignature *signature = [self.sessionManager methodSignatureForSelector:handleSelector];
        NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:signature];
        [invocation setSelector:handleSelector];
        [invocation setTarget:self.sessionManager];
        
        NSUInteger reasonValue = (NSUInteger)reason;
        [invocation setArgument:&reasonValue atIndex:2];
        [invocation setArgument:&_mockDefaults atIndex:3];
        [invocation invoke];
    }
    
    // Then: Audio should not be restarted
    XCTAssertEqual(self.mockAudio.restartCallCount, 0);
}

// MARK: - applyPlaybackPreferences(defaults:) Tests

- (void)testApplyPlaybackPreferencesEnablesSpeakerMode {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: Speaker mode is enabled in preferences
    [self.mockDefaults setBool:YES forKey:@"AudioSpeakerPhoneMode"];
    
    // When: Applying playback preferences
    SEL applySelector = NSSelectorFromString(@"applyPlaybackPreferencesWithDefaults:");
    if ([self.sessionManager respondsToSelector:applySelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:applySelector withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    
    // Then: Should complete without error
    // (AVAudioSession state can't be verified in unit tests)
    XCTAssertNotNil(self.sessionManager);
}

- (void)testApplyPlaybackPreferencesDisablesSpeakerMode {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: Speaker mode is disabled in preferences
    [self.mockDefaults setBool:NO forKey:@"AudioSpeakerPhoneMode"];
    
    // When: Applying playback preferences
    SEL applySelector = NSSelectorFromString(@"applyPlaybackPreferencesWithDefaults:");
    if ([self.sessionManager respondsToSelector:applySelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:applySelector withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    
    // Then: Should complete without error
    XCTAssertNotNil(self.sessionManager);
}

- (void)testApplyPlaybackPreferencesUsesDefaultWhenNotSet {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: No speaker mode preference set
    // (defaults to false/unset)
    
    // When: Applying playback preferences
    SEL applySelector = NSSelectorFromString(@"applyPlaybackPreferencesWithDefaults:");
    if ([self.sessionManager respondsToSelector:applySelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:applySelector withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    
    // Then: Should complete without error
    XCTAssertNotNil(self.sessionManager);
}

// MARK: - Integration Tests

- (void)testBindAndRefreshIntegration {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: Speaker mode enabled and audio not running
    [self.mockDefaults setBool:YES forKey:@"AudioSpeakerPhoneMode"];
    self.mockAudio.mockIsRunning = NO;
    
    // When: Binding and then refreshing
    SEL bindSelector = NSSelectorFromString(@"bindToMumbleKitAudio:defaults:");
    if ([self.sessionManager respondsToSelector:bindSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:bindSelector withObject:self.mockAudio withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    self.mockAudio.startCallCount = 0;  // Reset after bind
    
    SEL refreshSelector = NSSelectorFromString(@"refreshPlaybackChain");
    if ([self.sessionManager respondsToSelector:refreshSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:refreshSelector];
#pragma clang diagnostic pop
    }
    
    // Then: Audio should be started
    XCTAssertEqual(self.mockAudio.startCallCount, 1);
}

- (void)testRouteChangeAppliesPreferences {
    if (!self.sessionManager) {
        XCTFail(@"Could not access MUAudioSessionManager.shared");
        return;
    }
    
    // Given: Audio is bound with speaker mode enabled
    [self.mockDefaults setBool:YES forKey:@"AudioSpeakerPhoneMode"];
    self.mockAudio.mockIsRunning = YES;
    SEL bindSelector = NSSelectorFromString(@"bindToMumbleKitAudio:defaults:");
    if ([self.sessionManager respondsToSelector:bindSelector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        [self.sessionManager performSelector:bindSelector withObject:self.mockAudio withObject:self.mockDefaults];
#pragma clang diagnostic pop
    }
    
    // When: Route changes and preferences are updated
    [self.mockDefaults setBool:NO forKey:@"AudioSpeakerPhoneMode"];
    AVAudioSessionRouteChangeReason reason = AVAudioSessionRouteChangeReasonNewDeviceAvailable;
    SEL handleSelector = NSSelectorFromString(@"handleRouteChangeWithReason:defaults:");
    if ([self.sessionManager respondsToSelector:handleSelector]) {
        NSMethodSignature *signature = [self.sessionManager methodSignatureForSelector:handleSelector];
        NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:signature];
        [invocation setSelector:handleSelector];
        [invocation setTarget:self.sessionManager];
        
        NSUInteger reasonValue = (NSUInteger)reason;
        [invocation setArgument:&reasonValue atIndex:2];
        [invocation setArgument:&_mockDefaults atIndex:3];
        [invocation invoke];
    }
    
    // Then: Should apply new preferences and restart audio
    XCTAssertEqual(self.mockAudio.restartCallCount, 1);
}

@end

File: Tests/MUAudioSessionManagerTests.swift
import XCTest
import AVFoundation
@testable import Mumble

#if __has_include(<MumbleKit/MKAudio.h>)
import MumbleKit
#else
// Minimal stubs to allow the tests to compile in environments where
// the real MumbleKit headers are not available (such as CI runners).
@objc class MKAudio: NSObject {
    @objc static func sharedAudio() -> MKAudio {
        return MKAudio()
    }
    
    @objc var isRunning: Bool = false
    
    @objc func start() {
        isRunning = true
    }
    
    @objc func stop() {
        isRunning = false
    }
    
    @objc func restart() {
        stop()
        start()
    }
}
#endif

class MUAudioSessionManagerTests: XCTestCase {
    var sessionManager: MUAudioSessionManager!
    var mockDefaults: UserDefaults!
    
    override func setUp() {
        super.setUp()
        sessionManager = MUAudioSessionManager.shared
        // Use a separate suite for tests to avoid polluting user defaults
        mockDefaults = UserDefaults(suiteName: "MUAudioSessionManagerTests")!
        mockDefaults.removePersistentDomain(forName: "MUAudioSessionManagerTests")
    }
    
    override func tearDown() {
        mockDefaults.removePersistentDomain(forName: "MUAudioSessionManagerTests")
        super.tearDown()
    }
    
    // MARK: - bind(to:defaults:) Tests
    
    func testBindToMumbleKitAudioAppliesPlaybackPreferences() {
        // Given: Speaker mode is enabled in defaults
        mockDefaults.set(true, forKey: "AudioSpeakerPhoneMode")
        
        let audio = MKAudio.sharedAudio()
        
        // When: Binding to the audio instance
        sessionManager.bind(to: audio, defaults: mockDefaults)
        
        // Then: The playback preferences should be applied
        // We can verify this indirectly by checking that the method completes without error
        XCTAssertTrue(true, "bind method should complete without throwing")
    }
    
    func testBindWithSpeakerModeDisabled() {
        // Given: Speaker mode is disabled in defaults
        mockDefaults.set(false, forKey: "AudioSpeakerPhoneMode")
        
        let audio = MKAudio.sharedAudio()
        
        // When: Binding to the audio instance
        sessionManager.bind(to: audio, defaults: mockDefaults)
        
        // Then: The method should complete successfully
        XCTAssertTrue(true, "bind method should handle disabled speaker mode")
    }
    
    // MARK: - refreshPlaybackChain() Tests
    
    func testRefreshPlaybackChainCompletesSuccessfully() {
        // Given: A session manager with bound audio
        let audio = MKAudio.sharedAudio()
        sessionManager.bind(to: audio, defaults: mockDefaults)
        
        // When: Refreshing the playback chain
        sessionManager.refreshPlaybackChain()
        
        // Then: The method should complete without error
        XCTAssertTrue(true, "refreshPlaybackChain should complete without throwing")
    }
    
    // MARK: - handleRouteChange(reasonValue:defaults:) Tests
    
    func testHandleRouteChangeWithNewDeviceAvailable() {
        // Given: A route change with new device available
        let audio = MKAudio.sharedAudio()
        audio.start()
        sessionManager.bind(to: audio, defaults: mockDefaults)
        
        // When: Handling a route change for new device available
        let reason = AVAudioSession.RouteChangeReason.newDeviceAvailable
        sessionManager.handleRouteChange(reasonValue: reason.rawValue, defaults: mockDefaults)
        
        // Then: The audio subsystem should be restarted
        // This is verified by the method completing without error
        XCTAssertTrue(true, "handleRouteChange should handle new device available")
    }
    
    func testHandleRouteChangeWithOldDeviceUnavailable() {
        // Given: A route change with old device unavailable
        let audio = MKAudio.sharedAudio()
        audio.start()
        sessionManager.bind(to: audio, defaults: mockDefaults)
        
        // When: Handling a route change for old device unavailable
        let reason = AVAudioSession.RouteChangeReason.oldDeviceUnavailable
        sessionManager.handleRouteChange(reasonValue: reason.rawValue, defaults: mockDefaults)
        
        // Then: The audio subsystem should be restarted
        XCTAssertTrue(true, "handleRouteChange should handle old device unavailable")
    }
    
    func testHandleRouteChangeWithCategoryChange() {
        // Given: A route change with category change
        let audio = MKAudio.sharedAudio()
        audio.start()
        sessionManager.bind(to: audio, defaults: mockDefaults)
        
        // When: Handling a route change for category change
        let reason = AVAudioSession.RouteChangeReason.categoryChange
        sessionManager.handleRouteChange(reasonValue: reason.rawValue, defaults: mockDefaults)
        
        // Then: The audio subsystem should be restarted
        XCTAssertTrue(true, "handleRouteChange should handle category change")
    }
    
    func testHandleRouteChangeWithUnknownReason() {
        // Given: A route change with unknown reason
        let audio = MKAudio.sharedAudio()
        sessionManager.bind(to: audio, defaults: mockDefaults)
        
        // When: Handling a route change for unknown reason
        let reason = AVAudioSession.RouteChangeReason.unknown
        sessionManager.handleRouteChange(reasonValue: reason.rawValue, defaults: mockDefaults)
        
        // Then: The method should complete without restarting audio
        XCTAssertTrue(true, "handleRouteChange should handle unknown reason gracefully")
    }
    
    func testHandleRouteChangeWithOverride() {
        // Given: A route change with override reason
        let audio = MKAudio.sharedAudio()
        audio.start()
        sessionManager.bind(to: audio, defaults: mockDefaults)
        
        // When: Handling a route change for override
        let reason = AVAudioSession.RouteChangeReason.override
        sessionManager.handleRouteChange(reasonValue: reason.rawValue, defaults: mockDefaults)
        
        // Then: The audio subsystem should be restarted
        XCTAssertTrue(true, "handleRouteChange should handle override")
    }
    
    // MARK: - applyPlaybackPreferences(defaults:) Tests
    
    func testApplyPlaybackPreferencesWithSpeakerModeEnabled() {
        // Given: Speaker mode is enabled in defaults
        mockDefaults.set(true, forKey: "AudioSpeakerPhoneMode")
        
        // When: Applying playback preferences
        sessionManager.applyPlaybackPreferences(defaults: mockDefaults)
        
        // Then: The method should complete successfully
        XCTAssertTrue(true, "applyPlaybackPreferences should apply speaker mode")
    }
    
    func testApplyPlaybackPreferencesWithSpeakerModeDisabled() {
        // Given: Speaker mode is disabled in defaults
        mockDefaults.set(false, forKey: "AudioSpeakerPhoneMode")
        
        // When: Applying playback preferences
        sessionManager.applyPlaybackPreferences(defaults: mockDefaults)
        
        // Then: The method should complete successfully
        XCTAssertTrue(true, "applyPlaybackPreferences should apply receiver mode")
    }
    
    func testApplyPlaybackPreferencesWithDefaultValue() {
        // Given: No explicit speaker mode setting in defaults (should default to false)
        // When: Applying playback preferences
        sessionManager.applyPlaybackPreferences(defaults: mockDefaults)
        
        // Then: The method should complete successfully with default value
        XCTAssertTrue(true, "applyPlaybackPreferences should handle default value")
    }
    
    // MARK: - configureSession(activate:) Tests
    
    func testConfigureSessionWithActivation() {
        // Given: A session manager
        // When: Configuring the session with activation
        sessionManager.configureSession(activate: true)
        
        // Then: The session should be configured and activated
        XCTAssertTrue(true, "configureSession should activate when requested")
    }
    
    func testConfigureSessionWithoutActivation() {
        // Given: A session manager
        // When: Configuring the session without activation
        sessionManager.configureSession(activate: false)
        
        // Then: The session should be configured but not activated
        XCTAssertTrue(true, "configureSession should not activate when not requested")
    }
    
    func testConfigureSessionDefaultActivation() {
        // Given: A session manager
        // When: Configuring the session with default parameter
        sessionManager.configureSession()
        
        // Then: The session should be activated by default
        XCTAssertTrue(true, "configureSession should activate by default")
    }
}
